kong/runloop/balancer/targets.lua

```diff
@@ -176,14 +177,23 @@ function targets_M.on_target_event(operation, target)
     return
   end
 
-  -- cancel DNS renewal
-  if operation ~= "create" then
-    local key, err = get_dns_renewal_key(target)
-    if key then
-      renewal_weak_cache[key] = nil
-      renewal_heap:remove(key)
-    else
-      log(ERR, "could not stop DNS renewal for target removed from ", upstream_id, ": ", err)
+  local function cancel_dns_renewal(target_entity, op)
+    if op ~= "create" then
+      local key, err = get_dns_renewal_key(target_entity)
+      if key then
+        renewal_weak_cache[key] = nil
+        renewal_heap:remove(key)
+        return true
+      else
+        log(ERR, "could not stop DNS renewal for target ", target_entity, " removed from ", upstream_id, ": ", err)
+        return false
+      end
+    end
+  end
+
+  if not cancel_dns_renewal(target, operation) then
+    for _, t in ipairs(targets_list) do
```

Comment:
When the `operation` is equal to `create`, doesn't that mean we don't need to do the iteration here? If so, wouldn't it be better for `cancel_dns_renewal` to return a `true` by default?

Reply:
Good catch!